source/extensions/access_loggers/wasm/config.cc

```diff
@@ -20,30 +20,29 @@ namespace AccessLoggers {
 namespace Wasm {
 
 AccessLog::InstanceSharedPtr
-WasmAccessLogFactory::createAccessLogInstance(const Protobuf::Message& config,
+WasmAccessLogFactory::createAccessLogInstance(const Protobuf::Message& proto_config,
                                               AccessLog::FilterPtr&& filter,
                                               Server::Configuration::FactoryContext& context) {
-  const auto& wal_config =
-      MessageUtil::downcastAndValidate<const envoy::config::accesslog::v2::WasmAccessLog&>(config);
+  const auto& config =
+      MessageUtil::downcastAndValidate<const envoy::config::accesslog::v2::WasmAccessLog&>(
+          proto_config);
+  auto id = config.id();
+  auto configuration = std::make_shared<std::string>(config.configuration());
   auto tls_slot = context.threadLocal().allocateSlot();
-  if (wal_config.has_vm_config()) {
-    auto base_wasm =
-        Common::Wasm::createWasm(wal_config.id(), wal_config.vm_config(), context.api());
-    tls_slot->set([&wal_config, &base_wasm, &context](Event::Dispatcher& dispatcher) {
-      auto result =
-          Common::Wasm::createThreadLocalWasm(*base_wasm, wal_config.vm_config(), dispatcher,
-                                              wal_config.configuration(), context.api());
-      result->setClusterManager(context.clusterManager());
+  if (config.has_vm_config()) {
+    auto base_wasm = Common::Wasm::createWasm(id, config.vm_config(), context.clusterManager(),
+                                              context.dispatcher(), context.api());
+    // NB: the Slot set() call doesn't complete inline, so all arguments must outlive this call.
+    tls_slot->set([base_wasm, configuration](Event::Dispatcher& dispatcher) {
+      auto result = Common::Wasm::createThreadLocalWasm(*base_wasm, *configuration, dispatcher);
       return std::static_pointer_cast<ThreadLocal::ThreadLocalObject>(result);
     });
   } else {
-    if (wal_config.id().empty()) {
+    if (id.empty()) {
       throw Common::Wasm::WasmVmException("No WASM VM Id or vm_config specified");
     }
-    tls_slot->set([&wal_config, &context](Event::Dispatcher& dispatcher) {
-      auto result = Common::Wasm::getThreadLocalWasm(wal_config.id(), wal_config.configuration());
-      result->setDispatcher(dispatcher);
-      result->setClusterManager(context.clusterManager());
+    tls_slot->set([id, configuration](Event::Dispatcher&) {
```

Comment:
configuration = std::move(configuration)

Reply:
Generally I don't do optimizations like this that are off the critical path and outside of loops.   If they add any complexity I feel they are a net loss in the supportability vs performance game. 

If you really want it done I'll do it but otherwise I'd rather favor simplicity.

Reply:
No worry. I don't think it worth blocking the merge. 
Sorry so far I don't have the knowledge to determine whether it is critical path.

Reply:
These VMs are setup on startup and configuration change pushes.  The process is relatively expensive already, so a couple atomics is not going to break the bank.